### PR TITLE
fix Issue 14036 - Do not throw FinalizeError on OutOfMemoryError or Inva...

### PR DESCRIPTION
**Pull request by CyberShadow**
...lidMemoryOperationError

This changes FinalizeError to be thrown only when an Exception is thrown,
with the disadvantage of not catching non-Exception Throwables and the
advantage of preserving the stack trace for the code that caused
InvalidMemoryOperationError.

https://issues.dlang.org/show_bug.cgi?id=14036

### CODE REVIEW

**Comment by CyberShadow** [comment]
@MartinNowak

File: src/rt/lifetime.d [hunk 1]
```diff
@@ -2591,33 +2591,75 @@ unittest
...
+        {
+            caught = true;
+        }
+        GC.free(cast(void*)c);
+        return caught;
+    }
+
+    assert( test!Exception);
+    import core.exception : InvalidMemoryOperationError;
+    assert(!test!InvalidMemoryOperationError);
```
Comment (rainers) [inline src/rt/lifetime.d]: I guess you'll have to catch the memory error to successfully pass the unittest.

**Comment by rainers** [comment]
I agree that not catching Errors is better. 
Slightly related: we should use some special assert inside the GC to avoid an InvalidMemoryError because onAssertError is allocating.

**Comment by CyberShadow** [comment]
Finally fixed tests.

**Comment by MartinNowak** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
Why was the stacktrace of Throwable.next missing?

**Comment by CyberShadow** [comment]
We can allocate neither a stack trace nor a FinalizeError in an InvalidMemoryOperation condition.

**Comment by CyberShadow** [comment]
To clarify, I was talking about the stack trace in the debugger. If we rethrow, the stack trace of the point where the exception was thrown is lost (unless it was saved in the exception object, which it was not).

**Comment by Orvid** [comment]
With an InvalidMemoryError or OutOfMemoryError, would it be a viable choice to simply claim that all memory currently allocated is free to be used by the runtime to output useful information, such as a stack trace. Both of these errors are impossible to recover from, so the best we can do is hope that the current state is at least valid enough to be able to use the space for a stack trace.

**Comment by CyberShadow** [comment]
That might do more harm than good, e.g. corrupt user data.

**Comment by CyberShadow** [comment]
It would also be easier to use C malloc, or allocate directly from the OS.

**Comment by Orvid** [comment]
A short transcript of the ensuing IRC conversation:

<Orvid> CyberShadow: In both of those cases though, there's nothing you can do about it, the program IS going to crash, so why does it matter if we corrupt user data?
<CyberShadow> The user might be dumb and set up some atexit() handlers OSLT even though they shouldn't
<Orvid> But for it to corrupt their data, they'd have to be using GC allocated memory; As we are already known to be crashing, the entirety of the application's memory should be assumed to be corrupted, we are merely being wishful that the GC's core tables haven't been.

**Comment by MartinNowak** [comment]
A stack trace for OOM is misleading and InvalidMemoryError will be fixed at some point.
We could probably still use malloc for the stacktrace.

**Pull request merged**